README.md
# PyANU
• PyANU is a data science software application, packed with features for co-piloting and calculating various outcomes for remotelyoperated
underwater vehicles.
• Features include: estimating and plotting a plane’s crash site on a map, calculating distances using coordinate pixel conversion, color
and shape recognition, and various other tools.

# Changelog
v2.0.0:
*Implemented the GUI using PyQt4.
*Changed the wind equation to be variable in plotting the Crashmap.

v2.2.1:
*Added functionanility to the Seismo tab.

v2.2.2:
*Fixed the Seismo tabs switching in a random fashion.

v2.3.0:
*Changed the Horizontal function to measure on the mooring line.
*Added automatic screenshotting using PyScreenshot.
*Combined both functions above into the Horizontal tab.

v2.3.1:
*Added the NewCapture flag to know if the horizontal should make a new capture or use the last taken photo.
*Fixed the CSR by adding green, orange and yellow to be displayed as yellow.
*Added showing big letters instead of descriptive text on the screen.
*Removed the RES from the CSR.

v2.3.2:
*Adjusted CSR HSV color ranges to fit darker places via the Color Picker algorithm.
*Changed the wind equation to integrate over time using SciPy instead of updated incrementing to increase the accuracy.
*Changed many of the output printing methods in the Crashmap function.

v2.4.0:
*Added Real-Time OCR using Tesseract, OpenCV and PyScreenshot

v2.4.1:
*Added conditional aircraft type output to the OCR function.
*Improved OCR accuracy using multiple data pre-processing techniques.




Changelog.txt
v2.0.0:
*Implemented the GUI using PyQt4.
*Changed the wind equation to be variable in plotting the Crashmap.

v2.2.1:
*Added functionanility to the Seismo tab.

v2.2.2:
*Fixed the Seismo tabs switching in a random fashion.

v2.3.0:
*Changed the Horizontal function to measure on the mooring line.
*Added automatic screenshotting using PyScreenshot.
*Combined both functions above into the Horizontal tab.

v2.3.1:
*Added the NewCapture flag to know if the horizontal should make a new capture or use the last taken photo.
*Fixed the CSR by adding green, orange and yellow to be displayed as yellow.
*Added showing big letters instead of descriptive text on the screen.
*Removed the RES from the CSR.

v2.3.2:
*Adjusted CSR HSV color ranges to fit darker places via the Color Picker algorithm.
*Changed the wind equation to integrate over time using SciPy instead of updated incrementing to increase the accuracy.
*Changed many of the output printing methods in the Crashmap function.

v2.4.0:
*Added Real-Time OCR using Tesseract, OpenCV and PyScreenshot

v2.4.1:
*Added conditional aircraft type output to the OCR function.
*Improved OCR accuracy using multiple data pre-processing techniques.

PyANU.py
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import math
import numpy as np
import scipy.integrate
import sys

from PIL import Image
import pyscreenshot as ImageGrab

NewCapture = True #USED FOR SELECTING WHETHER THE HORIZONTAL FUNCTION WILL
                  #SCREENSHOT A NEW IMAGE OR USE THE LAST EXISTING ONE.

class PyANU:

    def Crashmap(self, takeoff, heading, vASC, RoC, vDES, RoD, fail, windAng, windSpd):
        img = mpimg.imread('Lake Washington.PNG')
        imgplot = plt.imshow(img)

        #heading, vASC, RoC, vDES, RoD, fail, windAng = 184, 93, 10, 64, 6, 43, 270  # RoC: RATE OF CLIMB, RoD: OF DECAY

        x_arr = []
        y_arr = []

        #print('\nTaking off from "Naval" or "Renton"?')
        #takeoff = input()

        if takeoff == 'Naval':
            x_arr.append(282.128)
            y_arr.append(761.84)
        elif takeoff == 'Renton':
            x_arr.append(350.805)
            y_arr.append(131.42)

        plt.ylim(0, 1040)  # INVERTING Y-AXIS WITHOUT INVERTING THE PIC

        # ----------------------------------------------------------------------------------------------------------------------
        # ----------------------------------------------------------------------------------------------------------------------

        # ASCENSION using vGS = cos(θ)∗vTAS+vwind
        print('\nASCENT\n')
        theta = math.degrees(math.atan(RoC / vASC))  # ANGLE OF ASCENDING
        vTAS = math.sqrt((RoC ** 2) + (vASC ** 2))  # VELOCITY OF TRUE AIR SPEED
        vGS = math.cos(theta * math.pi / 180) * vTAS  # VELOCITY OF GROUND SPEED
        dA = vGS * fail  # DISTANCE ASCENDED
        dAP = dA / 32.8  # DISTANCE IN PIXELS

        print('Plane covered ' + str(dA) + ' metres in ' + str(heading) + '° while ascending from take-off point')
        #print('\nCoordinates on X: ' + str(x_arr[0] + math.sin(heading * math.pi / 180) * dAP))
        #print('Coordinates on Y: ' + str(y_arr[0] + math.cos(heading * math.pi / 180) * dAP))

        print('\nMoved on X: ' + str(math.sin(heading * math.pi / 180) * dA) + ' metres')
        print('Moved on Y: ' + str(math.cos(heading * math.pi / 180) * dA) + ' metres')

        x_arr.append(x_arr[0] + math.sin(heading * math.pi / 180) * dAP)
        y_arr.append(y_arr[0] + math.cos(heading * math.pi / 180) * dAP)

        # ----------------------------------------------------------------------------------------------------------------------
        # ----------------------------------------------------------------------------------------------------------------------

        # DESCENT using vGS = cos(θ)∗vTAS+vwind
        print('\nDESCENT\n')
        vW = 0
        height = RoC * fail
        fall = height / RoD  # 71 secs
        theta = math.degrees(math.atan(RoD / vDES))  # ANGLE OF ASCENDING
        vTAS = math.sqrt((RoD ** 2) + (vDES ** 2))  # VELOCITY OF TRUE AIR SPEED
        vGS = math.cos(theta * math.pi / 180) * (vTAS + vW)  # VELOCITY OF GROUND SPEED
        dD = vGS * fall  # DISTANCE ASCENDED
        dDP = dD / 32.8  # DISTANCE IN PIXELS

        print('Plane covered ' + str(dD) + ' metres in ' + str(heading) + '° while descending from engine fail point')
        #print('\nCoordinates on X: ' + str(x_arr[1] + math.sin(heading * math.pi / 180) * dDP))
        #print('Coordinates on Y: ' + str(y_arr[1] + math.cos(heading * math.pi / 180) * dDP))

        print('\nMoved on X: ' + str(math.sin(heading * math.pi / 180) * dD) + ' metres')
        print('Moved on Y: ' + str(math.cos(heading * math.pi / 180) * dD) + ' metres')

        # ----------------------------------------------------------------------------------------------------------------------
        # ----------------------------------------------------------------------------------------------------------------------

        # WIND MOVEMENT
        print('\nWIND MOVEMENT\n')
        #for t in range(int(fall + 1)):
        #    vW += eval(windSpd)

        f = lambda t:eval(windSpd)
        vW = scipy.integrate.quad(f, 0, fall)

        print("\nvW: " + str(vW))

        vX = vW[0]
        vY = vW[1]

        vWX = vW[0] / 32.8 #X ON THE WORLD MAP IN PIXELS
        vWY = vW[1] / 32.8 #Y ON THE WORLD MAP IN PIXELS
        
        if (vY < 0.5): #FOR THE 270 WIND ANGLE SPECIAL CASE
            vY = 0

        print('On X: ' + str(vX) + ' metres' '\nOn Y: ' + str(vY) + ' metres')

        # ----------------------------------------------------------------------------------------------------------------------
        # ----------------------------------------------------------------------------------------------------------------------

        #TOTAL MOVEMENT
        print("\nTotal Movement: ")
        totalX = vX + (math.sin(heading * math.pi / 180) * dD) + (math.sin(heading * math.pi / 180) * dA)
        totalY = vY + (math.cos(heading * math.pi / 180) * dD) + (math.cos(heading * math.pi / 180) * dA)
        print("On X: " + str(totalX))
        print("On Y: " + str(totalY))
        print("\nReported Search Zone: " + str(((totalX**2)+(totalY**2))**0.5) + " meters, in direction: "
              + str(math.degrees(math.atan(totalX/totalY)) + 180) + " degrees from " + str(takeoff) + " point.")

        x_arr.append(x_arr[1] + (math.sin(heading * math.pi / 180) * dDP) + vWX)
        y_arr.append(y_arr[1] + (math.cos(heading * math.pi / 180) * dDP) + vWY)

        plt.plot(x_arr, y_arr, 'ro-')
        plt.show(imgplot)

    def Power(self, N, p, a, v, Cp):
        a = (float(a) ** 2) * 3.14159
        v = float(v) * 0.51444
        Cp = float(Cp) / 100

        P = (N * 0.5 * p * a * (v ** 3) * Cp) / (10 ** 6)
        print('Expected Power = ' + str(format(P, '.2f')) + 'MW')
        return str(format(P, '.2f'))

    def Vertical(self, p, P2, P1):
        '''P = ρgh'''
        h = (P2 - P1) / p * 9.81
        print('Distance between the two points= ' + str(format(h, '.2f')) + ' Metre(s)')
        return str(format(h, '.2f'))

    def Horizontal(self, ob_dis):
        if NewCapture is True:
            capture = ImageGrab.grab(bbox=(20, 80, 1050, 870))
            capture.save("Capture.png", "PNG")
        test_image = "Capture.png"
        img = mpimg.imread(test_image)
        NumberOfMouseClicks, x1, x2, y1, y2, distance_1, distance_2 = 0, 0, 0, 0, 0, 0, 0

        def onclick(event):
            x = event.xdata
            nonlocal NumberOfMouseClicks, x1, y1, x2, y2  #CHANGED FROM GLOBAL TO NONLOCAL
            print(event.xdata, event.ydata)

            #REFERENCE OBJECT
            if NumberOfMouseClicks == 0:
                x1 = x
            elif NumberOfMouseClicks == 1:
                y1 = x

            #STARTING POINT
            elif NumberOfMouseClicks == 2:
                x2 = x

            #UPDATE POINTS
            else:
                y2 = x
                distance_1 = y1 - x1
                distance_2 = y2 - x2
                #NumberOfMouseClicks = (-1)
                global objdist
                objdist = abs(ob_dis * distance_2 / distance_1)
                print('Unknown object distance = ', str(abs(ob_dis * distance_2 / distance_1)) + ' cm')
                #plt.close('all')
               
            NumberOfMouseClicks += 1

        fig, ax = plt.subplots()
        ax.plot(range(10))
        fig.canvas.mpl_connect('button_press_event', onclick)
        imgplot = plt.imshow(img)
        plt.show()
        plt.close()
        return format(objdist, '.2f')


def main():
    ROV = PyANU()
    #ROV.Crashmap()
    #ROV.Power()
    #ROV.Seismo()
    #ROV.Vertical()
    #ROV.Horizontal()


if __name__ == '__main__':
        main()


Extra/PyANUConsole.py
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import math
import numpy as np


class PyROV:

    def Crashmap(self):
        img = mpimg.imread('Lake Washington.PNG')
        imgplot = plt.imshow(img)

        heading, vASC, RoC, vDES, RoD, fail, windAng = 184, 93, 10, 64, 6, 43, 270  # RoC: RATE OF CLIMB, RoD: OF DECAY

        x_arr = []
        y_arr = []

        print('\nTaking off from "Naval" or "Renton"?')
        takeoff = input()

        if takeoff == 'Naval':
            x_arr.append(282.128)
            y_arr.append(761.84)
        elif takeoff == 'Renton':
            x_arr.append(350.805)
            y_arr.append(131.42)

        plt.ylim(0, 1040)  # INVERTING Y-AXIS WITHOUT INVERTING THE PIC

        # ASCENSION using vGS = cos(θ)∗vTAS+vwind
        print('\nASCENT\n')
        theta = math.degrees(math.atan(RoC / vASC))  # ANGLE OF ASCENDING
        vTAS = math.sqrt((RoC ** 2) + (vASC ** 2))  # VELOCITY OF TRUE AIR SPEED
        vGS = math.cos(theta * math.pi / 180) * vTAS  # VELOCITY OF GROUND SPEED
        dA = vGS * fail  # DISTANCE ASCENDED
        dAP = dA / 32.8  # DISTANCE IN PIXELS

        print('Plane covered ' + str(dA) + ' metres in ' + str(heading) + '° while ascending from take-off point')
        print('\nCoordinates on X: ' + str(x_arr[0] + math.sin(heading * math.pi / 180) * dAP))
        print('Coordinates on Y: ' + str(y_arr[0] + math.cos(heading * math.pi / 180) * dAP))

        print('\nMoved on X: ' + str(math.sin(heading * math.pi / 180) * dA) + ' metres and ' +
              str(math.sin(heading * math.pi / 180) * dAP) + ' pixels')
        print('Moved on Y: ' + str(math.cos(heading * math.pi / 180) * dA) + ' metres and ' +
              str(math.cos(heading * math.pi / 180) * dAP) + ' pixels')

        x_arr.append(x_arr[0] + math.sin(heading * math.pi / 180) * dAP)
        y_arr.append(y_arr[0] + math.cos(heading * math.pi / 180) * dAP)

        # ----------------------------------------------------------------------------------------------------------------------
        # ----------------------------------------------------------------------------------------------------------------------

        # DESCENT using vGS = cos(θ)∗vTAS+vwind
        print('\nDESCENT\n')
        vW = 0
        height = RoC * fail
        fall = height / RoD  # 71 secs
        theta = math.degrees(math.atan(RoD / vDES))  # ANGLE OF ASCENDING
        vTAS = math.sqrt((RoD ** 2) + (vDES ** 2))  # VELOCITY OF TRUE AIR SPEED
        vGS = math.cos(theta * math.pi / 180) * (vTAS + vW)  # VELOCITY OF GROUND SPEED
        dD = vGS * fall  # DISTANCE ASCENDED
        dDP = dD / 32.8  # DISTANCE IN PIXELS

        print('Plane covered ' + str(dD) + ' metres in ' + str(heading) + '° while descending from engine fail point')
        print('\nCoordinates on X: ' + str(x_arr[1] + math.sin(heading * math.pi / 180) * dDP))
        print('Coordinates on Y: ' + str(y_arr[1] + math.cos(heading * math.pi / 180) * dDP))

        print('\nMoved on X: ' + str(math.sin(heading * math.pi / 180) * dD) + ' metres and ' +
              str(math.sin(heading * math.pi / 180) * dDP) + ' pixels')
        print('Moved on Y: ' + str(math.cos(heading * math.pi / 180) * dD) + ' metres and ' +
              str(math.cos(heading * math.pi / 180) * dDP) + ' pixels')

        # WIND MOVEMENT
        print('\nWIND MOVEMENT\n')
        for i in range(int(fall + 1)):
            vW += -((1 / 720) * (i ** 2)) + 25

        vWP = vW / 32.8
        vWX = vWP * abs(math.sin(windAng * math.pi / 180))
        vWY = vWP * abs(math.cos(windAng * math.pi / 180))

        print('WIND COVERED ' + str(vW) + ' metres. On X: ' + str(vWX * 32.8) + ' metres & ' + str(
            vWX) + ' pixels, on Y: ' +
              str(vWY * 32.8) + ' metres & ' + str(vWY) + ' pixels.')

        x_arr.append(x_arr[1] + (math.sin(heading * math.pi / 180) * dDP) + vWX)
        y_arr.append(y_arr[1] + (math.cos(heading * math.pi / 180) * dDP) + vWY)

        plt.plot(x_arr, y_arr, 'ro-')
        plt.show(imgplot)

    def Seismo(self):
        '''x = np.array([0, 1.1, 1.3, 1.4, 1.5, 2, 2.2, 2.4, 2.6, 3, 3.1, 3.2, 3.3, 3.9])
        y = np.array([0, 0, 4.4, -4.3, 0, 0, 6.8, -7, 0, 0, 1.6, -1.6, 0, 0])'''
        x, y = [0]*16, [0]*16
        for i in range(16):
            data = input()
            datasplit = data.split()
            x[i] = float(datasplit[0])
            y[i] = float(datasplit[1])
        print(data)
        plt.xlabel('Time(s)')
        plt.ylabel('Amplitude')
        plt.plot(x, y, 'bo-')
        plt.grid(color='grey', linestyle='-', linewidth=0.5)
        plt.show()

    def Power(self):
        P, N, p, a, v, Cp = 0, 0, 0, 0, 0, 0

        print('\nInsert No. of Turbines')
        N = float(input())
        print('Insert the Density of the Water in kg/m^3')
        p = float(input())
        print('Insert the Radius of a Turbine in Metres')
        a = (float(input()) ** 2) * 3.14159
        print('Insert the Tide Strength in Knots')
        v = float(input()) * 0.51444
        print('Insert the Efficiency of the Turbines in %')
        Cp = float(input()) / 100

        P = (N * 0.5 * p * a * (v ** 3) * Cp) / (10 ** 6)
        print('Expected Power = ' + str(format(P, '.2f')) + 'MW')

    def Vertical(self):
        P1, P2, p, h = 0, 0, 0, 0
        print('\nInsert the Density of the Water in kg/m^3 at 20°C')
        p = float(input())
        print('Insert the Pressure at the Bottom point in Pascal')
        P2 = float(input())
        print('Insert the Pressure at the Top point in Pascal')
        P1 = float(input())

        '''P = ρgh'''
        h = (P2 - P1) / p * 9.81
        print('Distance between the two points= ' + str(format(h, '.2f')) + ' Metre(s)')

    def Horizontal(self):
        test_image = "Lake Washington.PNG"
        img = mpimg.imread(test_image)
        print('Enter known object distance in cm: ')
        ob_dis = int(input())
        NumberOfMouseClicks, x1, x2, y1, y2, distance_1, distance_2 = 0, 0, 0, 0, 0, 0, 0

        def onclick(event):
            x = event.xdata
            nonlocal NumberOfMouseClicks, x1, y1, x2, y2  #CHANGED FROM GLOBAL TO NONLOCAL
            print(event.xdata, event.ydata)
            if NumberOfMouseClicks == 0:
                x1 = x
            elif NumberOfMouseClicks == 1:
                y1 = x
            elif NumberOfMouseClicks == 2:
                x2 = x
            elif NumberOfMouseClicks == 3:
                y2 = x
                distance_1 = y1 - x1
                distance_2 = y2 - x2
                NumberOfMouseClicks = (-1)
                print('Unknown object distance = ', str(abs(ob_dis * distance_2 / distance_1)) + ' cm')
            NumberOfMouseClicks += 1

        fig, ax = plt.subplots()
        ax.plot(range(10))
        fig.canvas.mpl_connect('button_press_event', onclick)
        imgplot = plt.imshow(img)
        plt.show()


def main():
    ROV = PyROV()
    #ROV.Crashmap()
    ROV.Power()
    #ROV.Seismo()
    #ROV.Vertical()
    #ROV.Horizontal()


if __name__ == '__main__':
        main()


CSR.py
import numpy as np
import argparse
import imutils
import cv2

class ShapeDetector:
	def __init__(self):
		pass

	def detect(self, c):
		shape = "unidentified"
		peri = cv2.arcLength(c, True)
		approx = cv2.approxPolyDP(c, 0.04 * peri, True)

		if len(approx) == 3:
			shape = "triangle"

		elif len(approx) == 4:
			(x, y, w, h) = cv2.boundingRect(approx)
			ar = w / float(h)

			shape = "square" if ar >= 0.95 and ar <= 1.05 else "rectangle"

		elif len(approx) == 5:
			shape = "pentagon"

		#else:
		#	shape = "circle"

		return shape

class CSR:

    def RunCSR(self):
        ap = argparse.ArgumentParser()
        ap.add_argument("-v", "--video",
                        help="path to the (optional) video file")
        ap.add_argument("-b", "--buffer", type=int, default=64,
                        help="max buffer size")
        args = vars(ap.parse_args())

        #OLD RANGES
        '''lower = {'red': (166, 84, 141), 'green': (66, 122, 129), 'blue': (102, 104, 121), 'yellow': (23, 59, 119),
         'orange': (0, 50, 80)}  # assign new item lower['blue'] = (93, 10, 0)
        upper = {'red': (186, 255, 255), 'green': (86, 255, 255), 'blue': (119, 255, 255), 'yellow': (54, 255, 255),
         'orange': (20, 255, 255)}'''
        
        lower = {'red': (163, 84, 93), 'green': (66, 122, 129), 'blue': (100, 86, 8), 'yellow': (8, 133, 55),
         'orange': (0, 50, 80)}  # assign new item lower['blue'] = (93, 10, 0)
        upper = {'red': (187, 237, 255), 'green': (86, 255, 255), 'blue': (124, 237, 246), 'yellow': (33, 227, 266),
         'orange': (20, 255, 255)}

        colors = {'red': (0, 0, 255), 'green': (0, 255, 0), 'blue': (255, 2, 2), 'yellow': (0, 255, 217),
         'orange': (0, 140, 255)}

        if not args.get("video", False):
            camera = cv2.VideoCapture(0)

        else:
            camera = cv2.VideoCapture(args["video"])
            
        while True:
            (grabbed, frame) = camera.read()
            if args.get("video") and not grabbed:
                break

            #frame = imutils.resize(frame, width=600)

            #convert it to grayscale, and then blur it using a Gaussian filter with a 11 x 11 kernel.
            blurred = cv2.GaussianBlur(frame, (11, 11), 0)
            hsv = cv2.cvtColor(blurred, cv2.COLOR_BGR2HSV)
            sd = ShapeDetector()
            
            for key, value in upper.items():
                kernel = np.ones((9, 9), np.uint8)
                mask = cv2.inRange(hsv, lower[key], upper[key])
                mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
                mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel)
                cnts = cv2.findContours(mask.copy(), cv2.RETR_EXTERNAL,
                                        cv2.CHAIN_APPROX_SIMPLE)[-2]
                center = None


                if len(cnts) > 0:
                    c = max(cnts, key=cv2.contourArea)
                    ((x, y), radius) = cv2.minEnclosingCircle(c)
                    M = cv2.moments(c)
                    center = (int(M["m10"] / M["m00"]), int(M["m01"] / M["m00"]))
                    shape = sd.detect(c)
                    if radius > 0.5:
                        cv2.drawContours(frame, [c], -1, (0, 255, 0), 2)
                        
                        if(key == "red" and shape == "triangle"): 
                                cv2.putText(frame, "A", (int(x - radius), int(y - radius)), cv2.FONT_HERSHEY_SIMPLEX, 2.0,
                                        colors["red"], 2)
                        if((key == "yellow" or key == "orange" or key == "green") and shape == "triangle"): 
                                cv2.putText(frame, "B", (int(x - radius), int(y - radius)), cv2.FONT_HERSHEY_SIMPLEX, 2.0,
                                        colors["yellow"], 2)
                        if(key == "blue" and shape == "triangle"): 
                                cv2.putText(frame, "C", (int(x - radius), int(y - radius)), cv2.FONT_HERSHEY_SIMPLEX, 2.0,
                                        colors["blue"], 2)
                        if(key == "red" and shape == "rectangle"): 
                                cv2.putText(frame, "D", (int(x - radius), int(y - radius)), cv2.FONT_HERSHEY_SIMPLEX, 2.0,
                                        colors["red"], 2)
                        if((key == "yellow" or key == "orange" or key == "green") and shape == "rectangle"): 
                                cv2.putText(frame, "E", (int(x - radius), int(y - radius)), cv2.FONT_HERSHEY_SIMPLEX, 2.0,
                                        colors["yellow"], 2)
                        if(key == "blue" and shape == "rectangle"): 
                                cv2.putText(frame, "F", (int(x - radius), int(y - radius)), cv2.FONT_HERSHEY_SIMPLEX, 2.0,
                                        colors["blue"], 2)

            cv2.imshow("Frame", frame)
            key = cv2.waitKey(1) & 0xFF
            if key == ord("q"):
                break

        camera.release()
        cv2.destroyAllWindows()

if __name__ == '__main__':
    NCSR = CSR()
    NCSR.RunCSR()
